com.unity.render-pipelines.universal/ShaderLibrary/SSAO.hlsl

```diff
@@ -0,0 +1,452 @@
+TEXTURE2D_X(_BaseMap);
+TEXTURE2D_X(_ScreenSpaceAmbientOcclusionTexture);
+
+SAMPLER(sampler_BaseMap);
+SAMPLER(sampler_ScreenSpaceAmbientOcclusionTexture);
+
+// SSAO Settings
+int _SampleCount;
+half _Intensity;
+half _Radius;
+float _DownSample;
+
+#define RANDOM(f) GenerateHashedRandomFloat(f)
+#define SAMPLE_BASEMAP(uv)  SAMPLE_TEXTURE2D_X(_BaseMap, sampler_BaseMap, UnityStereoTransformScreenSpaceTex(uv));
+#define SCREEN_PARAMS GetScaledScreenParams()
+#define INTENSITY _Intensity
+#define RADIUS _Radius
+#define DOWNSAMPLE _DownSample
+
+#if !defined(SHADER_API_GLES)
+    #define SAMPLE_COUNT _SampleCount
+#else
+    // GLES2: In many cases, dynamic looping is not supported.
+    #define SAMPLE_COUNT 3
+#endif
+
+// --------
+// Options for further customization
+// --------
+
+// By default, a 5-tap Gaussian with the linear sampling technique is used
+// in the bilateral noise filter. It can be replaced with a 7-tap Gaussian
+// with adaptive sampling by enabling the macro below. Although the
+// differences are not noticeable in most cases, it may provide preferable
+// results with some special usage (e.g. NPR without textureing).
+//#define BLUR_HIGH_QUALITY
+
+// The SampleNormal function normalizes samples from G-buffer because
+// they're possibly unnormalized. We can eliminate this if it can be said
+// that there is no wrong shader that outputs unnormalized normals.
+// #define VALIDATE_NORMALS
+
+// Uniformly distributed points on a unit sphere
+#define FIX_SAMPLING_PATTERN
+
+// The constant below determines the contrast of occlusion. This allows
+// users to control over/under occlusion. At the moment, this is not exposed
+// to the editor because it's rarely useful.
+static const float kContrast = 0.6;
+
+// The constant below controls the geometry-awareness of the bilateral
+// filter. The higher value, the more sensitive it is.
+static const float kGeometryCoeff = -1.0;
+
+
+// The constants below are used in the AO estimator. Beta is mainly used
+// for suppressing self-shadowing noise, and Epsilon is used to prevent
+// calculation underflow. See the paper (Morgan 2011 http://goo.gl/2iz3P)
+// for further details of these constants.
+static const float kBeta = 0.002;
+#define EPSILON         1.0e-4
+
+
+
+// Gamma encoding (only needed in gamma lighting mode)
+half EncodeAO(half x)
+{
+    #if UNITY_COLORSPACE_GAMMA
+        return 1.0 - max(LinearToSRGB(1.0 - saturate(x)), 0.0);
+    #else
+        return x;
+    #endif
+}
+
+// Accessors for packed AO/normal buffer
+half4 PackAONormal(half ao, half3 n)
+{
+    return half4(ao, n * 0.5 + 0.5);
+}
+
+half GetPackedAO(half4 p)
+{
+    return p.r;
+}
+
+half3 GetPackedNormal(half4 p)
+{
+    return p.gba * 2.0 - 1.0;
+}
+
+// Trigonometric function utility
+float2 CosSin(float theta)
+{
+    float sn, cs;
+    sincos(theta, sn, cs);
+    return float2(cs, sn);
+}
+
+// Check if the camera is perspective.
+// (returns 1.0 when orthographic)
+float CheckPerspective(float x)
```

Comment:
IsPerspectiveProjection() in ShaderVariableFunctions.hlsl?

Reply:
Fixed!